Sources/X509/Verifier/PolicyBuilder.swift

```diff
@@ -0,0 +1,156 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the SwiftCertificates open source project
+//
+// Copyright (c) 2023 Apple Inc. and the SwiftCertificates project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftCertificates project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import SwiftASN1
+
+@resultBuilder
+public struct PolicyBuilder {
+    @inlinable
+    public static func buildBlock(_ components: some VerifierPolicy) -> some VerifierPolicy {
+        components
+    }
+    
+    @inlinable
+    public static func buildExpression(_ expression: some VerifierPolicy) -> some VerifierPolicy {
+        expression
+    }
+}
+
+
+// MARK: empty policy
+extension PolicyBuilder {
+    @usableFromInline
+    struct Empty: VerifierPolicy {
+        @inlinable
+        var verifyingCriticalExtensions: [SwiftASN1.ASN1ObjectIdentifier] { [] }
+        
+        @inlinable
+        init() {}
+        
+        @inlinable
+        mutating func chainMeetsPolicyRequirements(chain: UnverifiedCertificateChain) async -> PolicyEvaluationResult {
+            .meetsPolicy
+        }
+    }
+    
+    @inlinable
+    public static func buildBlock() -> some VerifierPolicy {
+        Empty()
+    }
+}
+
+// MARK: concatenated policies
+extension PolicyBuilder {
+    @usableFromInline
+    struct Tuple2<First: VerifierPolicy, Second: VerifierPolicy>: VerifierPolicy {
+        @usableFromInline
+        var first: First
+        
+        @usableFromInline
+        var second: Second
+        
+        @inlinable
+        init(first: First, second: Second) {
+            self.first = first
+            self.second = second
+        }
+        
+        @inlinable
+        var verifyingCriticalExtensions: [SwiftASN1.ASN1ObjectIdentifier] {
+            first.verifyingCriticalExtensions + second.verifyingCriticalExtensions
```

Comment:
This pattern may want us to re-evaluate the way we handle `verifyingCriticalExtensions`. Currently it's assumed that producing this is fairly cheap.

Reply:
hm... we can also cache the result automatically in `buildFinalResult` and return a new type which has a stored property for `verifyingCriticalExtensions`. WDYT?

Reply:
I think that's probably the best way to do that, actually!